src/plugins/shared_dictionaries_plugin/shared_dictionaries_plugin.hpp

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include <gtest/gtest_prod.h>
+#include "hyrise.hpp"
+#include "shared_dictionaries_plugin/plugin_settings/jaccard_index_threshold_setting.hpp"
+#include "storage/dictionary_segment.hpp"
+#include "storage/table.hpp"
+#include "types.hpp"
+#include "utils/abstract_plugin.hpp"
+#include "utils/singleton.hpp"
+
+namespace opossum {
+
+template <typename T>
+using SegmentChunkPair = std::pair<std::shared_ptr<DictionarySegment<T>>, std::shared_ptr<Chunk>>;
+
+/**
+ * The intention of this plugin is to save memory by using "dictionary sharing"
+ * while trying to not decrease the performance of Hyrise.
+ *
+ * Per default every dictionary encoded segment in Hyrise has its own dictionary.
+ * This plugin compares the dictionaries within a column for their similarity
+ * using the jaccard-index. If the jaccard-index is equal or higher than the
+ * specified threshold and additionally if the merging does not increase the width
+ * of the attribute vector, this plugin creates a shared dictionary. Then, the plugin
+ * replaces the dictionary segment with a new dictionary segment that has a shared
+ * dictionary.
+ *
+ * With the start of the plugin, the dictionary sharing compressor is automatically
+ * started for every table in the database.
+ */
+class SharedDictionariesPlugin : public AbstractPlugin {
+ public:
+  inline static const std::string LOG_NAME = "SharedDictionariesPlugin";
+
+  struct SharedDictionariesStats {
+    uint64_t total_bytes_saved = 0;
+    uint64_t total_previous_bytes = 0;
+    uint64_t modified_previous_bytes = 0;
+    uint32_t num_merged_dictionaries = 0;
+    uint32_t num_shared_dictionaries = 0;
+    uint32_t num_existing_merged_dictionaries = 0;
+    uint32_t num_existing_shared_dictionaries = 0;
+  };
+
+  SharedDictionariesPlugin();
+
+  std::string description() const final;
+
+  void start() final;
+
+  void stop() final;
+
+  SharedDictionariesStats stats = {};
+
+ private:
+  StorageManager& _storage_manager;
+  LogManager& _log_manager;
+
+  /*
+   * The following settings are currently hard coded for the plugin but could be made configurable:
+   *   - option to disable auto start of compression
+   *     The plugin currently starts the merging of dictionaries automatically once it is loaded.
+   *   - option to disable tracking of compression statistics
+   *     The plugin currently tracks memory usage statistics for merged data.
+   *   - option to select all tables / include tables / exclude tables
+   *     The plugin currently tries to merge dictionary segments of all available tables.
+   */
```

Comment:
I think we can remove that. Also, disabling the automatic start of compression should be difficult because it would need to happen before the Plugin is loaded when the settings do not exist yet, shouldn't it?

Reply:
We removed it. Fixed in https://github.com/hyrise/hyrise/pull/2394/commits/8ab006a666ed496b67ec0c88e833a1c96a269a3a